File: delivery_ops_bridge/adapters/feishu.py
from __future__ import annotations

import json
import subprocess
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Dict, List, Optional

from ..config import FeishuConfig
from ..models import Mention, SourceMessage, utc_now_iso

WORKING_REACTION_EMOJI_TYPE = "OnIt"


@dataclass
class SendResult:
    ok: bool
    raw: Dict[str, Any]
    chat_id: Optional[str] = None
    message_id: Optional[str] = None
    file_token: Optional[str] = None
    url: Optional[str] = None
    error: Optional[str] = None
    warning: Optional[str] = None


class FeishuEventParser:
    def __init__(self, bot_open_id: str, known_names: Dict[str, str] | None = None):
        self.bot_open_id = bot_open_id
        self.known_names = known_names or {}

    def parse(self, payload: Dict[str, Any]) -> Optional[SourceMessage]:
        event = payload.get("event", {})
        message = event.get("message", {})
        sender = event.get("sender", {})
        if not message:
            return None

        message_id = message.get("message_id") or message.get("message_id_v2") or message.get("root_id") or utc_now_iso()
        chat_id = message.get("chat_id", "")
        chat_type = message.get("chat_type", "private")
        message_type = message.get("message_type", "text")
        sender_open_id = sender.get("sender_id", {}).get("open_id", "")
        sender_name = self.known_names.get(sender_open_id, sender_open_id or "unknown")
        text = self._extract_text(message.get("content", ""))
        mentions = [self._parse_mention(item) for item in message.get("mentions", [])]
        mentions = [item for item in mentions if item.open_id]
        return SourceMessage(
            id=message_id,
            chat_id=chat_id,
            chat_type=chat_type,
            sender_open_id=sender_open_id,
            sender_name=sender_name,
            text=text,
            message_type=message_type,
            sent_at=message.get("create_time") or utc_now_iso(),
            raw_payload=payload,
            mentions=mentions,
            parent_id=message.get("parent_id"),
            root_id=message.get("root_id"),
        )

    def _extract_text(self, content: str) -> str:
        if not content:
            return ""
        try:
            parsed = json.loads(content)
        except json.JSONDecodeError:
            return content
        return parsed.get("text") or parsed.get("title") or content

    def _parse_mention(self, raw: Dict[str, Any]) -> Mention:
        mention_id = raw.get("id", {})
        open_id = raw.get("open_id") or mention_id.get("open_id") or raw.get("user_id", "")
        name = raw.get("name") or raw.get("key") or self.known_names.get(open_id, open_id)
        return Mention(open_id=open_id, name=name)


class FeishuAdapter:
    def __init__(self, config: FeishuConfig, dry_run: bool = False):
        self.config = config
        self.dry_run = dry_run

    def send_group_text(self, text: str, chat_id: str | None = None) -> SendResult:
        target = chat_id or self.config.group_chat_id
        return self._send(["--chat-id", target], text)

    def send_private_text(self, open_id: str, text: str) -> SendResult:
        return self._send(["--user-id", open_id], text)

    def send_reply_text(self, message_id: str, text: str) -> SendResult:
        if not message_id:
            return SendResult(ok=False, raw={}, error="message_id is required for reply")
        return self._send_reply(message_id, text)

    def upload_file(self, file_path: str) -> SendResult:
        if self.dry_run:
            return SendResult(ok=True, raw={"dry_run": True, "file_path": file_path}, url=file_path)
        path = Path(file_path).expanduser().resolve()
        try:
            proc = subprocess.run(
                [self.config.lark_cli_path, "drive", "+upload", "--as", "bot", "--file", f"./{path.name}"],
                check=False,
                capture_output=True,
                text=True,
                timeout=120,
                cwd=str(path.parent),
            )
        except Exception as exc:
            return SendResult(ok=False, raw={}, error=str(exc))
        return self._result_from_process(proc)

    def publish_file(self, file_path: str, share_link_entity: str = "tenant_readable") -> SendResult:
        upload = self.upload_file(file_path)
        if not upload.ok:
            return upload
        if self.dry_run or not upload.file_token:
            return upload

        permission = self._set_public_permission(upload.file_token, share_link_entity)
        if permission.ok:
            return SendResult(
                ok=True,
                raw={"upload": upload.raw, "permission": permission.raw},
                file_token=upload.file_token,
                url=upload.url,
            )

        return SendResult(
            ok=True,
            raw={"upload": upload.raw, "permission": permission.raw},
            file_token=upload.file_token,
            url=upload.url,
            warning=self._permission_warning(permission),
        )

    def _send(self, target_args: List[str], text: str) -> SendResult:
        if self.dry_run:
            raw = {"dry_run": True, "target": target_args, "text": text}
            return SendResult(ok=True, raw=raw, chat_id=None)
        cmd = [
            self.config.lark_cli_path,
            "im",
            "+messages-send",
            "--as",
            "bot",
            *target_args,
            "--msg-type",
            "text",
            "--text",
            text,
        ]
        try:
            proc = subprocess.run(cmd, check=False, capture_output=True, text=True, timeout=60)
        except Exception as exc:
            return SendResult(ok=False, raw={}, error=str(exc))
        return self._result_from_process(proc)

    def _send_reply(self, message_id: str, text: str) -> SendResult:
        if self.dry_run:
            raw = {"dry_run": True, "reply_to": message_id, "text": text}
            return SendResult(ok=True, raw=raw, chat_id=None)
        cmd = [
            self.config.lark_cli_path,
            "im",
            "+messages-reply",
            "--as",
            "bot",
            "--message-id",
            message_id,
            "--msg-type",
            "text",
            "--text",
            text,
        ]
        try:
            proc = subprocess.run(cmd, check=False, capture_output=True, text=True, timeout=60)
        except Exception as exc:
            return SendResult(ok=False, raw={}, error=str(exc))
        return self._result_from_process(proc)

    def add_reaction(self, message_id: str, emoji_type: str = WORKING_REACTION_EMOJI_TYPE) -> Optional[str]:
        if self.dry_run or not message_id:
            return None
        cmd = [
            self.config.lark_cli_path,
            "im", "reactions", "create",
            "--as", "bot",
            "--message-id", message_id,
            "--data", json.dumps({"reaction_type": {"emoji_type": emoji_type}}),
        ]
        try:
            proc = subprocess.run(cmd, check=False, capture_output=True, text=True, timeout=10)
            if proc.returncode == 0:
                raw = self._parse_json_output(proc.stdout)
                return raw.get("data", {}).get("reaction_id") or raw.get("reaction_id")
        except Exception:
            pass
        return None

    def remove_reaction(self, message_id: str, reaction_id: str) -> None:
        if self.dry_run or not message_id or not reaction_id:
            return
        cmd = [
            self.config.lark_cli_path,
            "im", "reactions", "delete",
            "--as", "bot",
            "--message-id", message_id,
            "--reaction-id", reaction_id,
        ]
        try:
            subprocess.run(cmd, check=False, capture_output=True, text=True, timeout=10)
        except Exception:
            pass

    def _set_public_permission(self, file_token: str, share_link_entity: str) -> SendResult:
        if self.dry_run:
            return SendResult(ok=True, raw={"dry_run": True, "file_token": file_token, "share_link_entity": share_link_entity})
        data = json.dumps(
            {
                "link_share_entity": share_link_entity,
                "external_access": False,
                "invite_external": False,
                "share_entity": "same_tenant",
                "security_entity": "anyone_can_view",
            },
            ensure_ascii=False,
        )
        cmd = [
            self.config.lark_cli_path,
            "drive",
            "permission.public",
            "patch",
            "--as",
            "bot",
            "--token",
            file_token,
            "--type",
            "file",
            "--data",
            data,
            "--yes",
        ]
        try:
            proc = subprocess.run(cmd, check=False, capture_output=True, text=True, timeout=60)
        except Exception as exc:
            return SendResult(ok=False, raw={}, error=str(exc))
        return self._result_from_process(proc)

    def _result_from_process(self, proc: subprocess.CompletedProcess[str]) -> SendResult:
        raw: Dict[str, Any] = {}
        if proc.stdout.strip():
            raw = self._parse_json_output(proc.stdout)
        chat_id, message_id, file_token, url = self._extract_result_fields(raw)
        if proc.returncode != 0:
            return SendResult(ok=False, raw=raw, chat_id=chat_id, message_id=message_id, file_token=file_token, url=url, error=proc.stderr.strip() or proc.stdout.strip())
        return SendResult(ok=True, raw=raw, chat_id=chat_id, message_id=message_id, file_token=file_token, url=url)

    def _parse_json_output(self, stdout: str) -> Dict[str, Any]:
        text = stdout.strip()
        try:
            return json.loads(text)
        except json.JSONDecodeError:
            pass

        start = text.find("{")
        if start >= 0:
            candidate = text[start:]
            try:
                return json.loads(candidate)
            except json.JSONDecodeError:
                pass
        return {"stdout": stdout}

    def _extract_result_fields(self, raw: Dict[str, Any]) -> tuple[Optional[str], Optional[str], Optional[str], Optional[str]]:
        scopes: List[Dict[str, Any]] = []
        data = raw.get("data")
        if isinstance(data, dict):
            scopes.append(data)
        scopes.append(raw)

        chat_id = self._first_value(scopes, ["chat_id"])
        message_id = self._first_value(scopes, ["message_id"])
        file_token = self._first_value(scopes, ["file_token", "token"])
        url = self._first_value(scopes, ["url", "file_url", "link"])
        return chat_id, message_id, file_token, url

    def _first_value(self, scopes: List[Dict[str, Any]], keys: List[str]) -> Optional[str]:
        for scope in scopes:
            for key in keys:
                value = scope.get(key)
                if isinstance(value, str) and value:
                    return value
            for nested_key in ("file", "item", "document"):
                nested = scope.get(nested_key)
                if isinstance(nested, dict):
                    nested_value = self._first_value([nested], keys)
                    if nested_value:
                        return nested_value
        return None

    def _permission_warning(self, result: SendResult) -> str:
        error = result.raw.get("error", {}) if isinstance(result.raw, dict) else {}
        missing_scopes = error.get("missing_scopes") or []
        if missing_scopes:
            scopes = "、".join(missing_scopes)
            return f"已上传看板文件，但未能设置共享权限。请在飞书开放平台为应用开通：{scopes}"
        return "已上传看板文件，但未能设置共享权限。"

    def explain_error(self, result: SendResult) -> str:
        error = result.raw.get("error", {}) if isinstance(result.raw, dict) else {}
        missing_scopes = error.get("missing_scopes") or []
        if missing_scopes:
            scopes = "、".join(missing_scopes)
            return f"请在飞书开放平台为应用开通以下权限：{scopes}"
        message = error.get("message")
        if message:
            return str(message)
        if result.error:
            return result.error.splitlines()[-1]
        return "未知错误"

File: backend/src/adapters/feishu.ts
import { execFile } from 'child_process';
import { promisify } from 'util';
import * as path from 'path';
import axios from 'axios';
import { FeishuConfig } from '../core/config';
import { SourceMessage, Mention, utcNowIso } from '../models/types';

const execFileAsync = promisify(execFile);

export const WORKING_REACTION_EMOJI_TYPES = ["Typing", "OnIt"];
export const WORKING_REACTION_EMOJI_TYPE = WORKING_REACTION_EMOJI_TYPES[0];

export interface SendResult {
    ok: boolean;
    raw: any;
    chat_id?: string;
    message_id?: string;
    file_token?: string;
    url?: string;
    error?: string;
    warning?: string;
}

export class FeishuEventParser {
    private botOpenId: string;
    private knownNames: Record<string, string>;

    constructor(botOpenId: string, knownNames: Record<string, string> = {}) {
        this.botOpenId = botOpenId;
        this.knownNames = knownNames;
    }

    parse(payload: any): SourceMessage | null {
        const event = payload?.event || {};
        const message = event.message || {};
        const sender = event.sender || {};
        if (Object.keys(message).length === 0) return null;

        const messageId = message.message_id || message.message_id_v2 || message.root_id || utcNowIso();
        const senderOpenId = sender.sender_id?.open_id || "";
        const senderName = this.knownNames[senderOpenId] || senderOpenId || "unknown";

        const mentionsRaw: any[] = message.mentions || [];
        const mentions: Mention[] = mentionsRaw
            .map(m => this._parseMention(m))
            .filter(m => !!m.open_id);

        return {
            id: messageId,
            chat_id: message.chat_id || "",
            chat_type: message.chat_type || "private",
            sender_open_id: senderOpenId,
            sender_name: senderName,
            text: this._extractText(message.content || ""),
            message_type: message.message_type || "text",
            sent_at: message.create_time || utcNowIso(),
            raw_payload: payload,
            mentions,
            parent_id: message.parent_id,
            root_id: message.root_id
        };
    }

    private _extractText(content: string): string {
        if (!content) return "";
        let parsed;
        try {
            parsed = JSON.parse(content);
        } catch {
            return content;
        }

        if (parsed.text) return parsed.text;

        if (parsed.content && Array.isArray(parsed.content)) {
            const lines: string[] = [];
            if (parsed.title) lines.push(parsed.title);
            for (const lineElements of parsed.content) {
                if (Array.isArray(lineElements)) {
                    let lineText = "";
                    for (const el of lineElements) {
                        if (el.tag === 'text' || el.tag === 'a') {
                            lineText += el.text || "";
                        } else if (el.tag === 'at') {
                            let name = el.user_name || el.name || "";
                            if (this.knownNames[el.user_id] && (!name || name.startsWith('_user_') || name === 'User')) {
                                name = this.knownNames[el.user_id];
                            } else if (!name || name.startsWith('_user_')) {
                                name = "未知成员";
                            }
                            lineText += `@${name} `;
                        }
                    }
                    lines.push(lineText);
                }
            }
            if (lines.length > 0) return lines.join('\n');
        }

        return parsed.title || content;
    }

    private _parseMention(raw: any): Mention {
        const mentionId = raw.id || {};
        const openId = raw.open_id || mentionId.open_id || raw.user_id || "";
        let name = raw.name || raw.key || "";
        if ((!name || name.startsWith("_user_")) && this.knownNames[openId]) {
            name = this.knownNames[openId];
        } else if (!name || name.startsWith("_user_")) {
            name = "未知成员";
        }
        return { open_id: openId, name };
    }
}

export class FeishuAdapter {
    private config: FeishuConfig;
    public dryRun: boolean;
    public lastReactionError: string | null = null;
    private auditCallback?: (eventType: string, payload: any) => void;
    private tokenCache: string | null = null;
    private tokenExpires: number = 0;

    constructor(config: FeishuConfig, dryRun: boolean = false) {
        this.config = config;
        this.dryRun = dryRun;
    }

    private async _getTenantAccessToken(): Promise<string> {
        if (this.tokenCache && this.tokenExpires > Date.now()) {
            return this.tokenCache;
        }

        const url = "https://open.feishu.cn/open-apis/auth/v3/tenant_access_token/internal";
        try {
            const resp = await axios.post(url, {
                app_id: this.config.app_id,
                app_secret: this.config.app_secret
            });
            if (resp.data.code === 0) {
                this.tokenCache = resp.data.tenant_access_token;
                this.tokenExpires = Date.now() + (resp.data.expire * 1000) - 60000;
                return this.tokenCache!;
            }
            throw new Error(`Token Error: ${JSON.stringify(resp.data)}`);
        } catch (e) {
            throw new Error(`Network error getting token: ${e}`);
        }
    }

    setAuditCallback(callback: (eventType: string, payload: any) => void) {
        this.auditCallback = callback;
    }

    async sendGroupText(text: string, chatId: string): Promise<SendResult> {
        const result = await this._send(["--chat-id", chatId], text);
        this._auditSend("group", chatId, text, result);
        return result;
    }

    async sendPrivateText(openId: string, text: string): Promise<SendResult> {
        const result = await this._send(["--user-id", openId], text);
        this._auditSend("private", openId, text, result);
        return result;
    }

    async sendReplyText(messageId: string, text: string): Promise<SendResult> {
        if (!messageId) {
            const res = { ok: false, raw: {}, error: "message_id is required for reply" };
            this._auditSend("reply", messageId, text, res);
            return res;
        }
        const result = await this._sendReply(messageId, text);
        this._auditSend("reply", messageId, text, result);
        return result;
    }

    async uploadFile(filePath: string): Promise<SendResult> {
        if (this.dryRun) return { ok: true, raw: { dry_run: true, filePath }, url: filePath };
        const resolvedPath = path.resolve(filePath);
        try {
            const proc = await this._runWithRetry(
                [this.config.lark_cli_path, "drive", "+upload", "--as", "bot", "--file", `./${path.basename(resolvedPath)}`],
                120000,
                path.dirname(resolvedPath)
            );
            return this._resultFromProcess(proc);
        } catch (e: any) {
            return { ok: false, raw: {}, error: String(e) };
        }
    }

    async publishFile(filePath: string, shareLinkEntity: string = "tenant_readable"): Promise<SendResult> {
        const upload = await this.uploadFile(filePath);
        if (!upload.ok || this.dryRun || !upload.file_token) return upload;

        const permission = await this._setPublicPermission(upload.file_token, shareLinkEntity);
        if (permission.ok) {
            return { ok: true, raw: { upload: upload.raw, permission: permission.raw }, file_token: upload.file_token, url: upload.url };
        }
        return { 
            ok: true, 
            raw: { upload: upload.raw, permission: permission.raw }, 
            file_token: upload.file_token, 
            url: upload.url, 
            warning: this._permissionWarning(permission) 
        };
    }

    private async _send(targetArgs: string[], text: string): Promise<SendResult> {
        if (this.dryRun) return { ok: true, raw: { dry_run: true, targetArgs, text }, chat_id: undefined };
        const cmd = [
            this.config.lark_cli_path, "im", "+messages-send", "--as", "bot", ...targetArgs, "--msg-type", "text", "--text", text
        ];
        try {
            const proc = await this._runWithRetry(cmd, 60000);
            return this._resultFromProcess(proc);
        } catch (e) {
            return { ok: false, raw: {}, error: String(e) };
        }
    }

    private async _sendReply(messageId: string, text: string): Promise<SendResult> {
        if (this.dryRun) return { ok: true, raw: { dry_run: true, messageId, text }, chat_id: undefined };
        const cmd = [
            this.config.lark_cli_path, "im", "+messages-reply", "--as", "bot", "--message-id", messageId, "--msg-type", "text", "--text", text
        ];
        try {
            const proc = await this._runWithRetry(cmd, 60000);
            return this._resultFromProcess(proc);
        } catch (e) {
            return { ok: false, raw: {}, error: String(e) };
        }
    }

    async addReaction(messageId: string, emojiType: string = WORKING_REACTION_EMOJI_TYPE): Promise<string | null> {
        this.lastReactionError = null;
        if (this.dryRun || !messageId) return null;

        try {
            const token = await this._getTenantAccessToken();
            const url = `https://open.feishu.cn/open-apis/im/v1/messages/${messageId}/reactions`;
            const resp = await axios.post(url, { reaction_type: { emoji_type: emojiType } }, {
                headers: { "Authorization": `Bearer ${token}` },
                timeout: 5000
            });
            if (resp.data.code === 0) {
                return resp.data.data?.reaction_id || null;
            }
            this.lastReactionError = `API returned: ${JSON.stringify(resp.data)}`;
        } catch (e) {
            this.lastReactionError = String(e);
        }
        return null;
    }

    async removeReaction(messageId: string, reactionId: string): Promise<void> {
        if (this.dryRun || !messageId || !reactionId) return;
        try {
            const token = await this._getTenantAccessToken();
            const url = `https://open.feishu.cn/open-apis/im/v1/messages/${messageId}/reactions/${reactionId}`;
            await axios.delete(url, {
                headers: { "Authorization": `Bearer ${token}` },
                timeout: 5000
            });
        } catch (e) {}
    }

    private async _setPublicPermission(fileToken: string, shareLinkEntity: string): Promise<SendResult> {
        if (this.dryRun) return { ok: true, raw: { dry_run: true, fileToken, shareLinkEntity } };
        const data = JSON.stringify({
            link_share_entity: shareLinkEntity,
            external_access: false,
            invite_external: false,
            share_entity: "same_tenant",
            security_entity: "anyone_can_view"
        });
        const cmd = [
            this.config.lark_cli_path, "drive", "permission.public", "patch", 
            "--as", "bot", "--token", fileToken, "--type", "file", "--data", data, "--yes"
        ];
        try {
            const proc = await this._runWithRetry(cmd, 60000);
            return this._resultFromProcess(proc);
        } catch (e) {
            return { ok: false, raw: {}, error: String(e) };
        }
    }

    private async _runWithRetry(cmd: string[], timeoutMs: number, cwd?: string): Promise<{stdout: string, stderr: string, code: number}> {
        const attempts = Math.max(1, (this.config.send_retry_count || 1) + 1);
        let lastResult = { stdout: "", stderr: "", code: 1 };
        
        for (let attempt = 0; attempt < attempts; attempt++) {
            try {
                const [executable, ...args] = cmd;
                const { stdout, stderr } = await execFileAsync(executable, args, { timeout: timeoutMs, cwd });
                return { stdout, stderr, code: 0 };
            } catch (e: any) {
                lastResult = { stdout: e.stdout || "", stderr: e.stderr || String(e), code: e.code || 1 };
                if (attempt === attempts - 1) return lastResult;
                await new Promise(r => setTimeout(r, Math.min(Math.pow(2, attempt) * 1000, 5000)));
            }
        }
        return lastResult;
    }

    private _resultFromProcess(proc: {stdout: string, stderr: string, code: number}): SendResult {
        let raw: any = {};
        if (proc.stdout.trim()) raw = this._parseJsonOutput(proc.stdout);
        if (Object.keys(raw).length === 0 && proc.stderr.trim()) raw = this._parseJsonOutput(proc.stderr);
        
        const { chatId, messageId, fileToken, url } = this._extractResultFields(raw);
        if (proc.code !== 0) {
            return { ok: false, raw, chat_id: chatId, message_id: messageId, file_token: fileToken, url, error: proc.stderr.trim() || proc.stdout.trim() };
        }
        return { ok: true, raw, chat_id: chatId, message_id: messageId, file_token: fileToken, url };
    }

    private _parseJsonOutput(stdout: string): any {
        const text = stdout.trim();
        try { return JSON.parse(text); } catch {}
        const start = text.indexOf('{');
        if (start >= 0) {
            try { return JSON.parse(text.slice(start)); } catch {}
        }
        return { stdout };
    }

    private _extractResultFields(raw: any) {
        const scopes: any[] = [];
        if (typeof raw?.data === 'object' && raw.data !== null) scopes.push(raw.data);
        scopes.push(raw);
        return {
            chatId: this._firstValue(scopes, ["chat_id"]),
            messageId: this._firstValue(scopes, ["message_id"]),
            fileToken: this._firstValue(scopes, ["file_token", "token"]),
            url: this._firstValue(scopes, ["url", "file_url", "link"])
        };
    }

    private _firstValue(scopes: any[], keys: string[]): string | undefined {
        for (const scope of scopes) {
            for (const key of keys) {
                if (typeof scope[key] === 'string' && scope[key]) return scope[key];
            }
            for (const nestedKey of ["file", "item", "document"]) {
                const nested = scope[nestedKey];
                if (typeof nested === 'object' && nested !== null) {
                    const nestedValue = this._firstValue([nested], keys);
                    if (nestedValue) return nestedValue;
                }
            }
        }
        return undefined;
    }

    private _auditSend(channel: string, target: string, text: string, result: SendResult) {
        if (!this.auditCallback) return;
        try {
            this.auditCallback("bot_message_sent", {
                channel, target, ok: result.ok, message_id: result.message_id,
                chat_id: result.chat_id, error: result.error, text_preview: text.substring(0, 200)
            });
        } catch {}
    }

    private _permissionWarning(result: SendResult): string {
        const error = result.raw?.error || {};
        const missing = error.missing_scopes || [];
        if (missing.length > 0) return `已上传看板文件，但未能设置共享权限。请在飞书开放平台为应用开通：${missing.join('、')}`;
        return "已上传看板文件，但未能设置共享权限。";
    }

    explainError(result: SendResult): string {
        const error = result.raw?.error || {};
        const missing = error.missing_scopes || [];
        if (missing.length > 0) return `请在飞书开放平台为应用开通以下权限：${missing.join('、')}`;
        if (error.message) return String(error.message);
        if (result.error) {
            const lines = result.error.split('\n');
            return lines[lines.length - 1];
        }
        return "未知错误";
    }

    parseEvent(payload: any): SourceMessage | null {
        const parser = new FeishuEventParser(this.config.bot_open_id || "");
        return parser.parse(payload);
    }

    async listGroups(): Promise<any[]> {
        if (this.dryRun) return [];
        try {
            const token = await this._getTenantAccessToken();
            const url = `https://open.feishu.cn/open-apis/im/v1/chats?page_size=100`;
            const resp = await axios.get(url, {
                headers: { "Authorization": `Bearer ${token}` }
            });
            return resp.data?.data?.items || [];
        } catch (e) {
            return [];
        }
    }

    async listGroupMembers(chatId: string): Promise<any[]> {
        if (this.dryRun || !chatId) return [];
        try {
            const token = await this._getTenantAccessToken();
            const url = `https://open.feishu.cn/open-apis/im/v1/chats/${chatId}/members?page_size=100`;
            const resp = await axios.get(url, {
                headers: { "Authorization": `Bearer ${token}` }
            });
            return resp.data?.data?.items || [];
        } catch (e) {
            return [];
        }
    }
}
